run_tests fails but the travis test won't fail unexpectedly (because of 0 exit code)

```
 



                                                                                
[jsk_perception:make] make[1]: *** [CMakeFiles/run_tests.dir/rule] Error 2




                                                                                
[jsk_perception:make] make[1]: Leaving directory `/workspace/ros/ws_jsk_recognition/build/jsk_perception'




                                                                                
[jsk_perception:make] make: *** [run_tests] Error 2



Failed <<< jsk_perception                         [ 0.5 seconds ]

Abandoned <<< jsk_pcl_ros                            [ Unrelated job failed ]

Abandoned <<< imagesift                              [ Unrelated job failed ]

Abandoned <<< resized_image_transport                [ Unrelated job failed ]


Finished <<< jsk_pcl_ros_utils                      [ 0.3 seconds ]

[build] Summary: 4 of 8 packages succeeded.

[build] Ignored: None.

[build] Warnings: None.

[build] Abandoned: 3 packages were abandoned.

[build] Failed: 1 packages failed.

[build] Runtime: 9.8 seconds total.

+ catkin_test_results build

Summary: 4 tests, 0 errors, 0 failures

+ travis_time_end

+ set +x

traivs_time:end:62ov6v3p:start=1461928060846147418,finish=1461928072949545664,duration=12103398246



traivs_fold:end:catkin_run_tests<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

Function catkin_run_tests takes 0 min 12 sec

+ '[' true '!=' true ']'

+ travis_time_start after_script

+ set +x

traivs_fold:start:after_script

traivs_time:start:oiqyfd5c>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

+ PATH=/usr/local/bin:/opt/ros/indigo/share/euslisp/jskeus/eus//Linux64/bin:/opt/ros/indigo/bin:/opt/ros/indigo/share/euslisp/jskeus/eus//Linux64/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin

+ PYTHONPATH=/usr/local/lib/python2.7/dist-packages:/workspace/ros/ws_jsk_recognition/devel/lib/python2.7/dist-packages:/opt/ros/indigo/lib/python2.7/dist-packages

+ '[' /workspace/jsk-ros-pkg/jsk_recognition/log == '' ']'

+ '[' catkin == catkin -a -e /workspace/jsk-ros-pkg/jsk_recognition/log ']'

+ catkin_test_results --verbose --all /workspace/jsk-ros-pkg/jsk_recognition/log

Summary: 0 tests, 0 errors, 0 failures

+ '[' catkin == catkin -a -e /workspace/ros/ws_jsk_recognition/build/ ']'

+ catkin_test_results --verbose --all /workspace/ros/ws_jsk_recognition/build/

Skipping "catkin_tools_prebuild/package.xml": 'tests'

jsk_recognition_utils/test_results/jsk_recognition_utils/gtest-test_cv_utils.xml: 1 tests

jsk_recognition_utils/test_results/jsk_recognition_utils/gtest-test_rgb_colors.xml: 1 tests

jsk_recognition_utils/test_results/jsk_recognition_utils/rostest-test_test_tf_listener_singleton.xml: 1 tests

jsk_recognition_utils/test_results/jsk_recognition_utils/rosunit-test_tf_listener_singleton.xml: 1 tests

Summary: 4 tests, 0 errors, 0 failures

+ '[' catkin == catkin -a -e /workspace/.ros/test_results/ ']'

+ travis_time_end

+ set +x

traivs_time:end:oiqyfd5c:start=1461928072952952892,finish=1461928073305733737,duration=352780845



traivs_fold:end:after_script<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

Function after_script takes 0 min 0 sec

+ pwd
/var/lib/jenkins/jobs/trusty-travis-jsk-ros-pkg-jsk_recognition-indigo-deb-true-true/workspace/jenkins-trusty-travis-jsk-ros-pkg-jsk_recognition-indigo-deb-true-true-1259/jsk-ros-pkg/jsk_recognition
+ sudo rm -fr /var/lib/jenkins/jobs/trusty-travis-jsk-ros-pkg-jsk_recognition-indigo-deb-true-true/workspace/jenkins-trusty-travis-jsk-ros-pkg-jsk_recognition-indigo-deb-true-true-1259
Finished: SUCCESS

=======================================
http://jenkins.jsk.imi.i.u-tokyo.ac.jp:8080/job/trusty-travis-jsk-ros-pkg-jsk_recognition-indigo-deb-true-true/1259/
=======================================
+++exit 0
```
https://travis-ci.org/jsk-ros-pkg/jsk_recognition/jobs/126623005

--- a/travis.sh
+++ b/travis.sh
@@ -203,7 +203,7 @@ fi
 if [ "$BUILDER" == catkin ]; then
     source devel/setup.bash > /tmp/$$.x 2>&1; grep export\ [^_] /tmp/$$.x ; rospack profile # force to update ROS_PACKAGE_PATH for rostest
     set -o pipefail  # this is necessary to pipe fail status on grepping
-    catkin run_tests -i --no-deps --no-status $TEST_PKGS $CATKIN_PARALLEL_TEST_JOBS --make-args $ROS_PARALLEL_TEST_JOBS | grep -v -e Symlinking -e Linked -e :[^\s]*install[^\s]*\]
+    catkin run_tests -i --no-deps --no-status $TEST_PKGS $CATKIN_PARALLEL_TEST_JOBS --make-args $ROS_PARALLEL_TEST_JOBS -- | grep -v -e Symlinking -e Linked -e :[^\s]*install[^\s]*\]
     set +o pipefail
     catkin_test_results build || error
 fi

--- a/travis.sh
+++ b/travis.sh
@@ -184,7 +184,11 @@ source /opt/ros/$ROS_DISTRO/setup.bash > /tmp/$$.x 2>&1; grep export\ [^_] /tmp/
 # for catkin
 if [ "${TARGET_PKGS// }" == "" ]; then export TARGET_PKGS=`catkin_topological_order ${CI_SOURCE_PATH} --only-names`; fi
 if [ "${TEST_PKGS// }" == "" ]; then export TEST_PKGS=$( [ "${BUILD_PKGS// }" == "" ] && echo "$TARGET_PKGS" || echo "$BUILD_PKGS"); fi
-if [ "$BUILDER" == catkin ]; then catkin build -i --summarize  --no-status $BUILD_PKGS $CATKIN_PARALLEL_JOBS --make-args $ROS_PARALLEL_JOBS | grep -v -e Symlinking -e Linked ; fi
+if [ "$BUILDER" == catkin ]; then
+  set -o pipefail  # this is necessary to pipe fail status on grepping
+  catkin build -i --summarize  --no-status $BUILD_PKGS $CATKIN_PARALLEL_JOBS --make-args $ROS_PARALLEL_JOBS | grep -v -e Symlinking -e Linked
+  set +o pipefail
+fi
 
 travis_time_end
 travis_time_start catkin_run_tests
@@ -198,7 +202,9 @@ fi
 
 if [ "$BUILDER" == catkin ]; then
     source devel/setup.bash > /tmp/$$.x 2>&1; grep export\ [^_] /tmp/$$.x ; rospack profile # force to update ROS_PACKAGE_PATH for rostest
+    set -o pipefail  # this is necessary to pipe fail status on grepping
     catkin run_tests -i --no-deps --no-status $TEST_PKGS $CATKIN_PARALLEL_TEST_JOBS --make-args $ROS_PARALLEL_TEST_JOBS | grep -v -e Symlinking -e Linked -e :[^\s]*install[^\s]*\]
+    set +o pipefail
     catkin_test_results build || error
 fi
 
@@ -211,7 +217,9 @@ if [ "$NOT_TEST_INSTALL" != "true" ]; then
     if [ "$BUILDER" == catkin ]; then
         catkin clean --yes || catkin clean -a # 0.3.1 uses -a, 0.4.0 uses --yes
         catkin config --install
+        set -o pipefail  # this is necessary to pipe fail status on grepping
         catkin build --summarize --no-status $BUILD_PKGS $CATKIN_PARALLEL_JOBS --make-args $ROS_PARALLEL_JOBS | grep -v -e Symlinking -e Linked -e :[^\s]*install[^\s]*\]
+        set +o pipefail
         source install/setup.bash > /tmp/$$.x 2>&1; grep export\ [^_] /tmp/$$.x
         rospack profile
         rospack plugins --attrib=plugin nodelet || echo "ok"

